### ISSUE TITLE
Document that UdpSocket::recv and recv_from do not read from the buffer

### ISSUE THREAD

**Issue by dtolnay** [carllerche/mio#657]
```rust
fn recv(&self, buf: &mut [u8]) -> Result<usize>;
fn recv_from(&self, buf: &mut [u8]) -> Result<(usize, SocketAddr)>;
```

It is important for these to document that they only write to the buffer, not read from it, because that means it is safe to pass an uninitialized buffer. It would be wasteful to construct a buffer, initialize it to contain all 0 bytes, and pass it to UdpSocket.

**Comment by frehberg** [comment]
And add to the docs, if recv returns with Ok(N), the following code must not use iterate over complete array, but iterate over a slice of the array buf[0,..,N-1].

**Comment by frehberg** [comment]
can be closed?

**Comment by dtolnay** [comment]
I don't think so. The docs in mio still need some work.

https://github.com/carllerche/mio/blob/5b9d20c4e01a63115ebdb6b87970fc9dd361a29b/src/net/udp.rs#L79-L81

**Comment by frehberg** [comment]
Finally the patch has been rebased, and could be merged

**Comment by carllerche** [comment]
The original PR got messed up somehow. I moved the commit to #775.

**Comment by carllerche** [comment]
Thanks. This has been merged (f6f293f0785d737fc7b18aac5c5f2b1368490d80).

**Comment by zheland** [comment]
The commit https://github.com/tokio-rs/mio/commit/f6f293f0785d737fc7b18aac5c5f2b1368490d80 updated the documentation for the wrong deprecated function that was doc-hidden and removed later. So the fix never reached the docs.

The commit updates docs for `recv_from` in `src/udp.rs` which was in:
```rust
#[deprecated(since = "0.6.6", note = "use net module instead")]
#[cfg(feature = "with-deprecated")]
#[doc(hidden)]
pub mod udp;
```
and removed in https://github.com/tokio-rs/mio/commit/e739c8a8a76a62e1939e24a0de6204d1720ee9a7 (remove deprecated code)

However, it is worth noting that the function by itself calls `std::net::UdpSocket` that doens't provide any "only write to the buffer" guarantees and, on the contrary, contain a requirement for "valid byte array" in docs:
```rust
    /// The function must be called with valid byte array `buf` of sufficient size to
    /// hold the message bytes. If a message is too long to fit in the supplied buffer,
    /// excess bytes may be discarded.
    /// ...
    pub fn recv_from(&self, buf: &mut [u8]) -> io::Result<(usize, SocketAddr)> {
```
